core/sile.lua

```diff
@@ -240,10 +241,251 @@ function SILE.resolveFile(filename, pathprefix)
+--- Must be popped with `pop(returnOfPush)`.
+function SILE.currentCommandStack:pushContent(content, tag, line, column, options, file)
+  return self:_pushFrame({
+    tag = tag or (type(content) == "table" and content.tag) or "???",
+    file = file or (type(content) == "table" and content.file) or SILE.currentlyProcessingFile,
+    line = line or (type(content) == "table" and content.line),
+    column = column or (type(content) == "table" and content.col),
+    options = options or (type(content) == "table" and content.attr) or {}
+  })
+end
+--- Push a text that is going to get typeset on to the stack to record the execution trace for debugging.
+--- Must be popped with `pop(returnOfPush)`.
+function SILE.currentCommandStack:pushText(text, line, column, file)
+  return self:_pushFrame({
+    text = text,
+    file = file,
+    line = line,
+    column = column
+  })
+end
+--- Internal: Push complete frame
+function SILE.currentCommandStack:_pushFrame(frame)
+  frame.toString = self._frameToString
+  frame.typesetter = SILE.typesetter
+  --frame.pushTrace = debug.traceback(nil, 3)
+
+  -- Push the frame
+  if SU.debugging("commandStack") then
+    print(string.rep(" ", #self) .. "PUSH(" .. frame:toString(false, true) .. ")")
+  end
+  self[#self + 1] = frame
+  self.lastCommand = nil
+  self._lastPushId = self._lastPushId + 1
+  frame.pushId = self._lastPushId
+  return self._lastPushId
+end
+--- Pop previously pushed command from the stack.
+--- Return value of `push` function must be provided as argument to check for balanced usage.
+function SILE.currentCommandStack:pop(pushId)
+  if type(pushId) ~= "number" then
+    SU.error("SILE.currentCommandStack:pop's argument must be the result value of the corresponding push", true)
+  end
+
+  -- First verify that push/pop is balanced
+  local popped = self[#self]
+  if popped.pushId ~= pushId then
+    local message = "Unbalanced content push/pop"
+    local debug = SILE.detailedErrors or SU.debugging("commandStack")
+    if debug then
+      message = message .. ". Expected " .. popped.pushId .. " - (" .. popped:toString() .. "), got " .. pushId
+    end
+    SU.warn(message, debug)
+  else
+    -- Correctly balanced: pop the frame
+    self.lastCommand = popped
+    self[#self] = nil
+    if SU.debugging("commandStack") then
+      print(string.rep(" ", #self) .. "POP(" .. popped:toString(false, true) .. ")")
+    end
+  end
+end
+--- Returns a stack of commands, which can be used to show information about location in document.
+--- Most relevant object is in the last index.
+--- Second return: recently popped value which could provide extra location information
+function SILE.currentCommandStack:locationStack(currentTypesetterOnly)
+  local result = {}
+
+  for i = 1, #self do
+    if not currentTypesetterOnly or (self[i].typesetter == SILE.typesetter) then
+      result[#result + 1] = self[i]
+    end
+  end
+
+  local extra = self.lastCommand
```

Comment:
This block is _really_ confusing. Don't test for something just to skip other tests, lets rearrange this entirely. Perhaps make it two tests, one `if currentTyesetterOnly and ...` and another for `if extra == ...`. But some of those checks seem to be done in previous blocks, so I'm really confused not just because the flow control here is weird but because it's unclear why it would be necessary at all.

Reply:
The second check is not needed anymore, as the `lastCommand` logic has changed, I'll remove that. The typesetter check is done a few lines earlier, but on different objects. It is still needed here, to properly filter the `extra` value.

That said, I have just now realized, that this whole method is no longer needed, except for filtering by typesetter. (Which probably seems obvious to you.) So the logic could be greatly simplified anyway. If the whole method isn't removed in the breadcrumb merge...

Reply:
I hope you don't feel like I'm nit-picking here. I realize we have quite a few dead code paths already in the repository, but these have grown up over time as things got refactored and dead code paths don't get caught right away. In a case like this of adding new code though I don't want to see stuff added that is effectively dead out of the gate.

Reply:
I agree with that, it was an oversight.